src/bin/ion/commands/jq.rs

```diff
@@ -78,34 +75,56 @@ fn evaluate_jq_expr(
     let modules = loader.load(&arena, program).unwrap();
 
     // compile the filter
-    let filter = jaq_core::Compiler::default()
+    jaq_core::Compiler::default()
         // Similar to `defs()` above, this would be our opportunity to extend the built-in filters
         .with_funs(jaq_std::funs::<JaqElement>())
         .compile(modules)
-        .unwrap();
+        .unwrap()
+}
 
-    // TODO: The setup above is happening once per input, but could probably happen once per run.
+fn evaluate_jq_expr(
+    input: CommandInput,
+    output: &mut CommandOutput,
+    filter: &Filter<Native<JaqElement>>,
+    slurp: bool,
+) -> anyhow::Result<()> {
 
     let mut reader = Reader::new(AnyEncoding, input.into_source())?;
-    let input_elements = reader.read_all_elements()?;
-    let ion_stream_as_element = List::from(input_elements).into();
+    let mut writer = output.as_writer()?;
+
+    if slurp {
+        let all_input_elements = reader.read_all_elements()?;
+        let slurped = List::from(all_input_elements).into();
+        filter_and_print(filter, &mut writer, slurped)?;
+    } else {
+        for item in reader.elements() {
+            let item: JaqElement = item?.into();
+            filter_and_print(filter, &mut writer, item)?;
+        }
+    }
+
+    writer.close()?;
+    Ok(())
+}
+
+fn filter_and_print(
+    filter: &Filter<Native<JaqElement>>,
+    writer: &mut CommandOutputWriter,
+    item: JaqElement
+) -> anyhow::Result<()> {
 
     let inputs = RcIter::new(core::iter::empty());
     // iterator over the output values
-    let out = filter.run((jaq_core::Ctx::new([], &inputs), ion_stream_as_element));
+    let ctx = Ctx::new([], &inputs); // can this be re-used? what even is this?
```

Comment:
It looks like the `Ctx` is used to store variables. I can't find any functionality that would allow you to carry a variable between filter runs, and their scope is defined to exist until the end of the filter. So, it doesn't seem like you'd want to re-use it, since it might pollute the next filter run.